File: greedymerge.py

s1 = "all is well"
s2 = "ell that en"
s3 = "hat end"
s4 = "t ends well"
s5 = "AAA"
s6 = "AAB"
s7 = "ABB"
s8 = "BBA"
s9 = "BBB"
s10 = "abxy"
s11 = "xyab"

fragementlist = [s1,s2,s3,s4, s5,s6,s7,s8,s9]

# iterate through fragment list until only one superstring is left 
# each iteration matches two fragments and then merges them into a single superstring
# if not two fragments match then concatenate the first two strings into a single superstring and continue 
while len(fragementlist) > 1:
    longestmatch = 0 # variable for determining which fragements have the most matching charachters in each loop
    matching_substring = '' # variable used to test substrings in each fragments
    merging_substring = '' # the substring that is contained within both merging fragements
    mergingfrag1 = '' # fragment 1 to merge 
    mergingfrag2 = '' # fragement 2 to merge

    # matching algorithm
    # iterate through each fragment in the fragment list and compare it char-by-char to the other fragments
    # contained within the list, determine which two fragments have the most matching charachters (in order!)
    # return the two fragements that have the most matching characthers to merge into the superstring
    for fragment1 in fragementlist:
        for fragment2 in fragementlist:
            substring = '' # variable to store the substring (used to determine matching fragments)
            count = 0 # variable to determine the amount of matching charachters between two fragments

            if fragment1 == fragment2: 
                continue # dont compare a fragment to itself
            else:
                # iterate through each characther in a fragment to determine matching substrings
                for letter in fragment1:
                    # construct a substring used to identify matching fragments
                    substring += letter
                    
                    # if the substring is in fragment 2 then increase the matching characther count by 1
                    if (substring in fragment2):
                        matching_substring = substring # update the matching substring 
                        count += 1
                    else:
                        continue
            
            # if two fragments have the most matching charachters then set them as the fragments to merge
            if count > longestmatch:
                longestmatch = count
                merging_substring = matching_substring
                mergingfrag1 = fragment1
                mergingfrag2 = fragment2

    # merging algorithm
    superstring = ""

    # ensure that merging fragmetns have a matching substring
    if (longestmatch != 0): 
        # determine the index or position within the fragment from which the match occurs
        index1 = mergingfrag1.find(merging_substring) # where substring is contained within first fragment
        index2 = mergingfrag2.find(merging_substring) # where substring is contained within the second fragment

        # if matching substring is the begining of the first fragment and NOT the second fragment
        if (index1 == 0)  & (index2 != 0):
            # add the 'leading' charachters of the second fragment before match occurs to the superstring
            superstring = mergingfrag2[0:index2]
            # add matching substring to merging superstring
            superstring += merging_substring

            # if there are additional characthers after matching substring then add them to the merging superstring
            if len(mergingfrag1) != len(merging_substring):
                superstring += mergingfrag1[len(merging_substring):len(mergingfrag1)]
        
        # if matching substring is the beginning of the second fragment and NOT the frist fragment
        elif (index1 != 0) & (index2 == 0):
            # add the 'leading' charachters of the second fragment before match occurs to the superstring
            superstring = mergingfrag1[0:index1]
            # add the matching substring to the merging superstring
            superstring += merging_substring

            # if trailing charachters in merging fragment then add them to superstring
            if len(mergingfrag2) != len(merging_substring):
                superstring += mergingfrag2[len(merging_substring):len(mergingfrag2)]
        
        # if substring is the beginning of BOTH fragments
        elif (index1 == 0) & (index2 == 0):
            
            # superstring will be the matching substring as it is the beginning of both fragments
            superstring = merging_substring

            # if fragment is larger than matching substring add additional charachters to superstring
            if (len(mergingfrag1) > len(merging_substring)):
                superstring += mergingfrag1[len(merging_substring):len(mergingfrag1)]
            
            # if fragment is larger than matching substring add additional charachters to superstring
            if (len(mergingfrag2) > len(merging_substring)):
                superstring += mergingfrag2[len(merging_substring):len(mergingfrag2)]
            

        fragementlist.append(superstring) # add merged fragments (superstring) to fragment list
        fragementlist.remove(mergingfrag1) # remove the merging fragment
        fragementlist.remove(mergingfrag2) # remove the mergning fragment

    # if not matching substring then start concatenating fragments
    else:
        # construct supertring out of the two 'front' fragments in the list
        superstring = fragementlist[0] + fragementlist[1]

        # remove the two merging fragments from the 'front' of the list
        fragementlist.remove(fragementlist[0])
        fragementlist.remove(fragementlist[0])

        # add merged fragments to list
        fragementlist.append(superstring)

    
# display final superstring
print(fragementlist[0])


    

File: GreedyMerge.cs
using System;
using System.Collections.Generic;

namespace GreedyMerge
{
    class Program
    {
        static void Main(string[] args)
        {
            String string1 = "all is well";
            String string2 = "ell that en";
            String string3 = "hat end";
            String string4 = "t ends well";
            String string5 = "AAA"; 
            String string6 = "AAB";
            String string7 = "ABB";
            String string8 = "BBA";
            String string9 = "BBB";

            List<string> fragmentList = new List<string>();
            fragmentList.Add(string1);
            fragmentList.Add(string2);
            fragmentList.Add(string3);
            fragmentList.Add(string4);
            fragmentList.Add(string5);
            fragmentList.Add(string6);
            fragmentList.Add(string7);
            fragmentList.Add(string8);
            fragmentList.Add(string9);


            // iterate through fragment list until only one superstring is left 
            // each iteration matches two fragments and then merges them into a single superstring
            //if not two fragments match then concatenate the first two strings into a single superstring and continue 
            while (fragmentList.Count != 1) {
                int longestmatch = 0; // variable for tracking which two fragments contain the most matches
                String matchingSubString = null; // variable for storing subStrings that match between fragment 1 & 2
                String mergingSubString = null; // variable for storing the subString of the two fragmetns being merged
                String mergingFragment1 = null; // variables for storingof the merging fragments
                String mergingFragment2 = null; 

                // matching algorithm
                // iterate through each fragment in the fragment list and compare it char-by-char to the other fragments
                // contained within the list, determine which two fragments have the most matching charachters (in order!)
                // return the two fragements that have the most matching characthers to merge into the superstring
                foreach (String fragment1 in fragmentList)
                {
                    foreach(String fragment2 in fragmentList)
                    {
                        String subString = ""; // variable for storing the substring and testing for matches between fragment 1 & 2
                        int count = 0; // variable for counting number of matching chars

                        if (fragment1 == fragment2) // dont compare fragments to themselves continue the loop
                        {
                            continue;
                        }
                        else
                        {
                            // iterate through each charachter in fragment1 to find matches with fragment2
                            foreach(Char letter in fragment1)
                            {
                                // build substring from fragment 1 letter by letter
                                subString += letter;

                                // if substring in fragment2 increase match count and store the substring in the matchingsubstring
                                if (fragment2.Contains(subString))
                                {
                                    count++;
                                    matchingSubString = subString;
                                } 
                                else // if substring not in fragment 2 go to the next fragment
                                {
                                    continue;
                                }
                            }
                        }

                        // determine if these two fragments have the most matching characthers
                        // if so store fragments and substring for merging and update longest match
                        if (count > longestmatch)
                        {
                            longestmatch = count;
                            mergingSubString = matchingSubString;
                            mergingFragment1 = fragment1;
                            mergingFragment2 = fragment2;
                        }
                    }
                }

                //merging algorithm 
                String superString = "";

                // ensure that fragments match to some extent. if not then concatenate
                if (longestmatch != 0){

                    int index1 = mergingFragment1.IndexOf(mergingSubString); // stores location of the substring in fragment 1
                    int index2 = mergingFragment2.IndexOf(mergingSubString); // stores location of the substring in fragment 2

                    int range1 = mergingFragment1.Length - mergingSubString.Length; // stores the amount of charachters in fragment 1 not in substring
                    int range2 = mergingFragment2.Length - mergingSubString.Length; // stores the amount of charachters in fragment 2 not in substring

                    // if substring is beginning of fragment 1 but not of fragment 2
                    if ((index1 == 0) && (index2 != 0))
                    {
                        // add the 'leading' portion of fragment 2 before the matching substring occurs
                        superString = mergingFragment2.Substring(0, index2);

                        // add the matching substring 
                        superString += mergingSubString;

                        // if fragment 1 contains trailing characthers not in the substring add them to the end of the superstring
                        if (mergingFragment1.Length != mergingSubString.Length)
                        {
                            superString += mergingFragment1.Substring(mergingSubString.Length, range1);
                        }

                    }
                    // if substring is the beginning of fragment 2 but not fragment 1
                    else if ((index1 != 0) && (index2 == 0))
                    {
                        // add the leading charachters before the substring from fragment 1
                        superString = mergingFragment1.Substring(0, index1);

                        // add the matching substring
                        superString += mergingSubString;

                        // if fragment 2 contains trailing charachters after the matching substring add them to the superstrinf
                        if (mergingFragment2.Length != mergingSubString.Length)
                        {
                            superString += mergingFragment2.Substring(mergingSubString.Length, range2);
                        }
                    } 
                    // if substring occurs at the beginning of BOTH fragment 1 and fragment 2
                    else if ((index1 == 0) && (index2 == 0))
                    {
                        // add the matching characthers to the superstring
                        superString = mergingSubString;

                        // if fragment 1 contains trailing charachters not in the matching substring add them to the end of superstring
                        if (mergingFragment1.Length > mergingSubString.Length)
                        {
                            superString += mergingFragment1.Substring(mergingSubString.Length, range1);
                        }
                        // if fragment 2 contains trailing charachters not in the matching substring add them to the end of superstring
                        if (mergingFragment2.Length > mergingSubString.Length)
                        {
                            superString += mergingFragment2.Substring(mergingSubString.Length, range2);
                        }

                    }

                    // after superstring is constructed 
                    // add superstring to the list and remove the fragments accordingly
                    fragmentList.Add(superString);
                    fragmentList.Remove(mergingFragment1);
                    fragmentList.Remove(mergingFragment2);
                }
                // if no match occurs between the fragments then concatenate the leading most fragments accordingly (the beginning of the list)
                else
                {
                    // concatenate the 2 most leading strings in list 
                    superString = fragmentList[0] + fragmentList[1];

                    // remove fragments from the list 
                    fragmentList.Remove(fragmentList[0]);
                    fragmentList.Remove(fragmentList[0]);

                    // add concatenated string to the list
                    fragmentList.Add(superString);
                }
            }

            // display the resulting message
            Console.WriteLine(fragmentList[0]);
        }
    }
}
